keps/sig-api-machinery/2896-openapi-v3/README.md

```diff
@@ -0,0 +1,523 @@
+
+### OpenAPI
+
+Swagger 2.0 and OpenAPI 3.0 have a couple of incompatible changes that are not currently supported by kube-openapi. Definitions are moved to components and the schema for Paths is slightly modified. A new struct and builder will be created in kube-openapi to represent the v3 schema to be imported by the necessary consumers.
+
+CRD structural schemas will be published as is without any value stripping for v3. Built-in types will generate the v3 OpenAPI definition, and either keep all fields or strip incompatible fields when building the swagger spec for v3 and v2 respectively.
+
+The OpenAPI handler will have an additional map to handle the publishing of v3.
+
+```go
+// Group -> OpenAPIv3 mapping
+var v3Schemas map[string]*OpenAPIv3
+
+type OpenAPIv3 struct {
+  rwMutex sync.RWMutex
+  lastModified time.Time
+
+  spec3Bytes []byte
+  spec3Pb []byte
+  spec3PbGz []byte
+
+  spec3BytesETag string
+  spec3PbETag string
+  spec3PbGzETag string
+}
+```
+
+### Version Skew
+
+There is a potential version skew between old aggregated apiservers and a new kube-apiserver. All new aggregated apiservers will publish v3 which will be discovered and published by the aggregator for v3. Old aggregated apiservers will not publish v3, and the aggregator cannot discover the v3 schema for the corresponding aggregated apiservers. The v3 schema for the old apiservers will not be published by the aggregator until they are upgraded to support v3. Regardless of version skew, v2 aggregation will continue for all aggregated apiservers.
+
+It is technically possible for the aggregator handling OpenAPI v3 to download from the OpenAPI v2 schema, but there are a few tradeoffs that I believe make this not worthwhile. The OpenAPI v3 aggregator must publish OpenAPI v3, and OpenAPI v2 will need to be converted if it is intended to be used. The v3 schema is designed to be a lossless representation of a user's spec, but by going through OpenAPI v2, critical information may have already been lost. Publishing an inaccurate v3 schema is not desirable and is not in line with the primary goal of this KEP. Converting from OpenAPI v2 to v3 via the Swagger/OpenAPI structs after publishing is also much more complicated than modifying the OpenAPI generation before publishing.
+
+By not publishing the v3 spec for old aggregated apiservers, the v3 spec is not a full representation of all the resources and endpoints available within the cluster. However, the scope of this is limited to select group-versions claimed by the select aggregated apiservers. A group-version in v3 will either have a fully lossless OpenAPI v3 or be non-existent (old aggregated apiserver), preventing a partial lossy spec from being published. This should also serve as a good incentive for users to upgrade their aggregated apiservers to the latest version and publish OpenAPI v3.
```

Comment:
I like the last sentence

Reply:
please split on new sentences so comments can be more effective.

> This should also serve as a good incentive for users to upgrade their aggregated apiservers to the latest version and publish OpenAPI v3.

This appears to conflate two different actors.  The people impacts are clients which are trying to get a view of all openapi information.  But the the people who have to update are the developers of aggregated apiservers and operators managing clusters.  This sentence implies that the users will benefit, but it seems equally likely to me that aggregated apiservers do not update and openapi v3 endpoints are unreliable and unusable for a large swatch of clients which cannot change the aggregated apiservers themselves.

If we want to update clients to use openapi v3, the information presented at the openapi v3 endpoints needs to include all the data they need.  Otherwise, it's strictly easier for clients (users) to never move, right?

Reply:
@sttts @apelisse I'd like to get your feedback on this since I'm not sure what the best way would be to balance the incentives. I agree with @deads2k's comment that it is technically still strictly easier for clients to never move to v3 because of potential incomplete information during the transition.

1. Current behavior: We publish v3 on a best effort basis. Clients consuming published v3 endpoints will benefit from the new v3 fields, and also experience a boost in performance (group-version split). However, clients would need to fall back to v2 if they need something from an aggregated apiserver that is not upgraded to v3. As @deads2k mentioned, depending on the cluster the v3 API may not be very usable until things are upgraded or the v2 fallback is fully ready. This provides an incentive to upgrade aggregated api servers to v3 but it may not be in the user's control. There needs to be coordination on upgrading everything around the same time to get the most benefit.

2. Alternative: For aggregated api servers, convert v2 to v3 (directly from spec, not from source go structs so the limitations of v2 still exist) and publish the "lossy" v3 schema for all aggregated api servers. This has the benefit that it'll be easier for clients to work with, but v3 is published as "lossy" and clients may see bugs that were supposed to be fixed by v3 remain until the aggregated servers are fully upgraded to v3. However, from the client's perspective things will appear to be working on v3, and I'm worried this may disincentivize aggregated apiservers from upgrading.

Reply:
> 2\. I'm worried this may disincentivize aggregated apiservers from upgrading.

If there is no openapi v2 to v3 conversion, I think that leaves the following client choices of 
1. speak both openapi v2 and openapi v3 and working with all kube-apiservers
2. speak just openapi v2 and work everywhere
3. speak just openapi v3 and work unpredictably

If you were a generic client (not trying to push this feature forward), which of those choices is easiest and does that encourage aggregated apiservers to move?

Reply:
Note that v2 -> v3 conversion will be even more lossy because we publish invalid specs today (openapi-gen is broken ever since and nobody fixed the specs). So we need heuristics to detect mistakes in the v2 spec from aggregated apiservers, assuming they got generated via openapi-gen. This will be messy. A requirement for v3 is that the new specs are finally correct.

Reply:
Realistically, we'll have to have a phase where both v3 and v2 are used to facilitate the skew.  I think that this is best done on the client side (use v3 but fallback to v2) so we have a clear distinction of which group/versions are v3 and which group/versions are v2.

As for a generic client, I'd say that "speaking just openapi v2 and work everywhere" is similar to speaking only openapi v3. v2 schemas are lossy and we already have invalid specs today so it already works somewhat unpredictably. If the client adopts supporting both v2 and v3, they'll be guaranteed to have full coverage and also lossless specs for a subset of the types. This is objectively better than the current behavior, although the direct benefit will depend on the # of old aggregated apiservers remaining.

Reply:
If openapi v3 is not complete, then I think we have to keep the openapi v2 indefinitely and it cannot be deprecated.  If we make openapi v3 complete, then I'm in favor of deprecating and eventually removing openapi v2.

Reply:
ack, added as beta graduation requirement